crypto/bio/bf_readbuff.c

```diff
@@ -0,0 +1,271 @@
+    BIO_F_BUFFER_CTX *ctx;
+    long ret = 1, sz;
+
+    ctx = (BIO_F_BUFFER_CTX *)b->ptr;
+
+    switch (cmd) {
+    case BIO_CTRL_EOF:
+        if (ctx->ibuf_len > 0)
+            return 0;
+        if (b->next_bio == NULL)
+            return 1;
+        ret = BIO_ctrl(b->next_bio, cmd, num, ptr);
+        break;
+
+    case BIO_C_FILE_SEEK:
+    case BIO_CTRL_RESET:
+        sz = ctx->ibuf_off + ctx->ibuf_len;
+        /* Assume it can only seek backwards */
+        if (num < 0 || num > sz)
+            return 0;
+        ctx->ibuf_off = num;
+        ctx->ibuf_len = sz - num;
+        break;
+
+    case BIO_C_FILE_TELL:
+    case BIO_CTRL_INFO:
+        ret = (long)ctx->ibuf_off;
+        break;
+    case BIO_CTRL_PENDING:
+        ret = (long)ctx->ibuf_len;
+        if (ret == 0) {
+            if (b->next_bio == NULL)
+                return 0;
+            ret = BIO_ctrl(b->next_bio, cmd, num, ptr);
+        }
+        break;
+    case BIO_CTRL_DUP:
+    case BIO_CTRL_FLUSH:
+        ret = 1;
+        break;
+    default:
+        ret = 0;
+        break;
+    }
+    return ret;
+}
+
+static long readbuffer_callback_ctrl(BIO *b, int cmd, BIO_info_cb *fp)
+{
+    if (b->next_bio == NULL)
+        return 0;
+    return BIO_callback_ctrl(b->next_bio, cmd, fp);
+}
+
+static int readbuffer_gets(BIO *b, char *buf, int size)
+{
+    BIO_F_BUFFER_CTX *ctx;
+    int num = 0, num_chars, found_newline;
+    char *p;
+
+    if (size == 0)
+        return 0;
+    --size; /* the passed in size includes the terminator - so remove it here */
+    ctx = (BIO_F_BUFFER_CTX *)b->ptr;
+    BIO_clear_retry_flags(b);
+
+    for (;;) {
+        if (ctx->ibuf_len > 0) {
+            p = &(ctx->ibuf[ctx->ibuf_off]);
+            found_newline = 0;
+            for (num_chars = 0;
+                 (num_chars < ctx->ibuf_len) && (num_chars < size);
+                 num_chars++) {
+                *(buf++) = p[num_chars];
+                if (p[num_chars] == '\n') {
+                    found_newline = 1;
+                    num_chars++;
+                    break;
+                }
+            }
+            num += num_chars;
+            size -= num_chars;
+            ctx->ibuf_len -= num_chars;
+            ctx->ibuf_off += num_chars;
+            if (found_newline || size == 0) {
+                *buf = '\0';
+                return num;
+            }
+        } else {
+            /* read another line and resize if we have to */
+            if (!readbuffer_resize(ctx, size))
+                return 0;
+
+            /* Read another line from the next bio using BIO_gets */
+            num_chars = BIO_gets(b->next_bio, ctx->ibuf + ctx->ibuf_off,
+                                 1 + size);
+            if (num_chars <= 0) {
+                BIO_copy_next_retry(b);
+                *buf = '\0';
+                if (num_chars < 0)
+                    return ((num > 0) ? num : num_chars);
+                else
+                    return num; /* i == 0 */
```

Comment:
Nit: The comment should be `num_chars == 0`, shouldn't it?

Also the whole if/else could be rewritten as `return num > 0 ? num : num_chars;`

Reply:
Change the comment : the ternary logic is different..

Reply:
IMO if you think it through you'll find that it isn't.

Reply:
There is an outer if that says num_chars <= 0 and the num >= 0 because it is just counter and cannot go negative (except for integer overflow but that is not handled anywhere anyway).  You return num in the else. But that is the same value as result of the `num > 0 ? num : num_chars` because num_chars is 0 in the else case and num is >= 0.

Reply:
I had to draw a table :)

Reply:
Note that I really expect you to care, but I hate `else` after `return`.  Extra parens on line 264. I'm also normally an advocate for usign `?:` but in this case it might make the code more complicated, rather than more clear.